### DISCUSSION TITLE
UserGroupsService method to add user to additional group(s)

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#9409] [Ideas]
> Created by: Evan Warner (evan@evanwarner.com) on 2016/06/09 19:36:44 +0000
> Votes at time of UserVoice import: 3

-----------------------------------

To programmatically add a user to a group, you need to use craft()->userGroups->assignUserToGroups(), but calling this will remove the user from any group they're in that was not passed to this call. This means you need to look up all the user's current groups and redundantly pass them to the call (which then inefficiently deletes the usergroups_users records and recreates them).

It would be really nice if there was a method that would simply add the user to the passed groups, and not worry about what groups they're already in.

Bonus points if this could be passed a list of either UserGroupModels or user group ids (assignUserToGroups requires ids).

**Comment by angrybrad** [comment]
> Posted by Lindsey DiLoreto (lindsey@doublesecretagency.com) on 2015/03/11 04:48:18 +0000
-----------------------------------

The group relationships would all get reset anyway... but I totally agree, it would be nice to have a method which doesn&#39;t require you to manually lookup the existing groups. +1

**Comment by lesaff** [comment]
I am doing exactly what @evanwarner was saying (checking for the user's existing group(s) and append the new group I want to assign it to. 

On my loop, `assignUserToGroups()` returns a `SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry` exception error and somehow resets that particular user's group.

Would it be better if the user already belongs to a certain group, and that group is in the new array, then it will just silently ignored it rather than being reset?

Code

```php
$today = (new \DateTime())->format(\DateTime::ATOM);

// Initialize user group array and set expired group as a default array entry
$expiredUserGroups = [12];

// Get all expired trial users
$expiredUsers = \craft\elements\User::find()
    ->group(['letterFreeTrialers', 'metalsFreeTrialers'])
    ->status(['active', 'locked', 'suspended', 'pending'])
    ->memberUntil("< {$today}")
    ->all();

// Loop through users...
foreach ($expiredUsers as $key => $user) {

    // Get user's current groups
    $currentGroups = $user->groups;

    // Create an array of user groups
    foreach ($currentGroups as $currentGroup) {
        array_push($expiredUserGroups, $currentGroup->id);
    }

    // Update user's profile and reset to expired groups accordingly
    $result = Craft::$app->getUsers()->assignUserToGroups($user->id, $expiredUserGroups);
}```

**Comment by brandonkelly** [comment]
@lesaff I think you need to move this line:

```php
$expiredUserGroups = [12];
```

into the `foreach` loop. Otherwise each additional user is going to keep appending user group IDs onto the existing array, leading to duplicates.

**Comment by lesaff** [comment]
@brandonkelly Good catch. That works. Thanks.

**Comment by michaelrog** [comment]
@brandonkelly would yall take a PR for `addUserToGroup` and `removeUserFromGroup` methods?

Perhaps w/ some convenience methods like `$user->addToGroup($group)` and `$group->removeUser($user)`?

**Comment by brandonkelly** [comment]
Sure why not.
